File: AI.py
# init(is_black)
# out: (i,j) the position of the chessboard
from const_set import *


import copy
import time
import random
from chessboard import ChessBoard




def evaluate(board,color):
	b_value,w_value = board.value_judge(color)
	if color == BLACK:

		return b_value - w_value
	else:
		return w_value - b_value


def evaluate_2(board,color,x,y): # for this step
	b_value,w_value = board.value_judge_2(x,y,color)
	if color == BLACK:

		return b_value - w_value
	else:
		return w_value - b_value


class MinMax():
	def __init__(self,is_black,depth = DEPTH):
		self.depth = depth
		if is_black == True:
			self.color = BLACK
			self.other = WHITE
		else:
			self.color = WHITE
			self.other = BLACK



	def get_move(self,board):
		self.over = False
		depth = self.depth
		self.count = 0
		start = time.time()
		value,idx = self.Max(board,None,depth,999999999)
		print("Time of this step: " + str(time.time() - start))
		print('AI_MINMAX had come a over:' + str(self.over))
		#print("The value is : " + str(value))
		return idx

	def Max(self,board,move,depth,cut_value):
		# self.count += 1
		# if self.count % 100 == 0:
		# 	print(self.count)
		my_board = copy.deepcopy(board)

		if move is not None:
			my_board.draw_xy(move[0],move[1],self.other)
			if my_board.anyone_win(move[0],move[1]) == self.color:
				self.over = True
				return 9999999 * (depth + 1),(move[0],move[1])
			if my_board.anyone_win(move[0],move[1]) == self.other:
				self.over = True
				return -9999999 * (depth + 1),(move[0],move[1])

		max_value = -999999999
		empty_item = my_board.get_k_dist_empty(3)
		if len(empty_item) == 0:
			empty_item.append([MIDDLE,MIDDLE])
		t = random.randint(0,len(empty_item)-1)
		max_idx = empty_item[t]
		depth -= 1
		#print('good location:  \n'+str(empty_item))
		#print('MAX_depth:'+str(depth))
		if depth < 0:
			return evaluate(my_board,self.color),(move[0],move[1])

		
		for mov in empty_item:
			
			value,idx = self.Min(my_board,mov,depth,max_value)
			if value > cut_value:
				return cut_value,(move[0],move[1])
			if max_value < value:
				max_value = value
				max_idx = (mov[0],mov[1])

		return max_value,max_idx





	def Min(self,board,move,depth,cut_value):
		# self.count += 1
		# if self.count % 100 == 0:
		# 	print(self.count)
		my_board = copy.deepcopy(board)
		if move is not None:
			my_board.draw_xy(move[0],move[1],self.color)
			if my_board.anyone_win(move[0],move[1]) == self.color:
				self.over = True
				return 9999999 * (depth + 1),(move[0],move[1])
			if my_board.anyone_win(move[0],move[1]) == self.other:
				self.over = True
				return -9999999 * (depth + 1),(move[0],move[1])

		min_value = 999999999
		empty_item = my_board.get_k_dist_empty(3)
		if len(empty_item) == 0:
			empty_item.append([MIDDLE,MIDDLE])
		t = random.randint(0,len(empty_item)-1)
		min_idx = empty_item[t]
		depth -= 1
		#print('MIN_depth:' + str(depth))
		if depth < 0:
			return evaluate(my_board,self.color),(move[0],move[1])


		for mov in empty_item:
			value,idx = self.Max(my_board,mov,depth,min_value)
			if value < cut_value:
				return cut_value,(move[0],move[1])

			if min_value > value:
				min_value = value
				min_idx = (mov[0],mov[1])

		return min_value,min_idx


# class MinMax_2():
# 	def __init__(self,is_black):
# 		if is_black == True:
# 			self.color = BLACK
# 			self.other = WHITE
# 		else:
# 			self.color = WHITE
# 			self.other = BLACK



# 	def get_move(self,board):
# 		self.over = False
# 		depth = DEPTH
# 		self.count = 0
# 		start = time.time()
# 		value,idx = self.Max(board,None,depth,9999999)
# 		print("Time of this step: " + str(time.time() - start))
# 		print("The value is : " + str(value))
# 		return idx

# 	def Max(self,board,move,depth,cut_value):
# 		# self.count += 1
# 		# if self.count % 1000 == 0:
# 		# 	print(self.count)
# 		my_board = copy.deepcopy(board)
# 		base_value = 0
# 		if move is not None:
# 			my_board.draw_xy(move[0],move[1],self.other)
# 			base_value = evaluate_2(my_board,self.color,move[0],move[1])
			
# 			if my_board.anyone_win(move[0],move[1]) == self.color:
# 				return 9999999,(move[0],move[1])
# 			if my_board.anyone_win(move[0],move[1]) == self.other:
# 				return -9999999,(move[0],move[1])
# 		empty_item,a,b = my_board.get_board_item()
# 		max_value = -9999999
# 		t = random.randint(0,len(empty_item)-1)
# 		max_idx = empty_item[t]
# 		depth -= 1
# 		#print('MAX_depth:'+str(depth))
# 		if depth < 0:
# 			return evaluate_2(my_board,self.color,move[0],move[1]),(move[0],move[1])

# 		if base_value > cut_value:
# 			return cut_value,(move[0],move[1])
# 		for mov in empty_item:
			
# 			v,idx = self.Min(my_board,mov,depth,max_value)
# 			value = base_value + v
# 			if value > cut_value:
# 				return cut_value,(move[0],move[1])
# 			if max_value < value:
# 				max_value = value
# 				max_idx = (mov[0],mov[1])
# 		return max_value,max_idx





	# def Min(self,board,move,depth,cut_value):
	# 	# self.count += 1
	# 	# if self.count % 100 == 0:
	# 	# 	print(self.count)
	# 	my_board = copy.deepcopy(board)
	# 	base_value = 0
	# 	if move is not None:
	# 		my_board.draw_xy(move[0],move[1],self.color)
	# 		base_value = evaluate_2(my_board,self.color,move[0],move[1])
			
	# 		if my_board.anyone_win(move[0],move[1]) == self.color:
	# 			return 9999999,(move[0],move[1])
	# 		if my_board.anyone_win(move[0],move[1]) == self.other:
	# 			return -9999999,(move[0],move[1])
	# 	empty_item,a,b = my_board.get_board_item()
	# 	min_value = 9999999
	# 	t = random.randint(0,len(empty_item)-1)
	# 	min_idx = empty_item[t]
	# 	depth -= 1
	# 	#print('MIN_depth:' + str(depth))
	# 	if depth < 0:
	# 		return evaluate_2(my_board,self.color,move[0],move[1]),(move[0],move[1])

	# 	if base_value < cut_value:
	# 		return cut_value,(move[0],move[1])
	# 	for mov in empty_item:
	# 		v,idx = self.Max(my_board,mov,depth,min_value)
	# 		value = base_value + v
	# 		if value < cut_value:
	# 			return cut_value,(move[0],move[1])

	# 		if min_value > value:
	# 			min_value = value
	# 			min_idx = (mov[0],mov[1])

	# 	return min_value,min_idx








class MinMax_smallset():
	def __init__(self,is_black,depth = DEPTH):
		self.depth = depth
		if is_black == True:
			self.color = BLACK
			self.other = WHITE
		else:
			self.color = WHITE
			self.other = BLACK



	def get_move(self,board):
		self.over = False
		depth = self.depth
		self.count = 0
		self.win_depth = 0
		start = time.time()
		value,idx = self.Max(board,None,depth,999999999)
		print("AI step: " + str(idx))
		print("Time of this step: " + str(time.time() - start))
		print('AI_MINMAX_SMALL had come a over:' + str(self.over))
		#print("The value is : " + str(value))
		return idx

	def Max(self,board,move,depth,cut_value):
		# self.count += 1
		# if self.count % 100 == 0:
		# 	print(self.count)
		my_board = copy.deepcopy(board)
		base_value = 0
		if move is not None:
			my_board.draw_xy(move[0],move[1],self.other)
			base_value = evaluate_2(my_board,self.color,move[0],move[1])
			
			if my_board.anyone_win(move[0],move[1]) == self.color:
				self.over = True
				return 9999999 * (depth + 1),(move[0],move[1])
			if my_board.anyone_win(move[0],move[1]) == self.other:
				self.over = True
				return -9999999 * (depth + 1),(move[0],move[1])
		empty_item = my_board.get_k_dist_empty(3)
		if len(empty_item) == 0:
			empty_item.append([MIDDLE,MIDDLE])
		max_value = -999999999

		t = random.randint(0,len(empty_item)-1)
		max_idx = empty_item[t]
		depth -= 1
		#print('MAX_depth:'+str(depth))
		if depth < 0:
			return evaluate_2(my_board,self.color,move[0],move[1]),(move[0],move[1])

		# if base_value > cut_value:
		# 	return cut_value,(move[0],move[1])
		for mov in empty_item:
			
			v,idx = self.Min(my_board,mov,depth,max_value)
			value = base_value + v
			if value > cut_value:
				return cut_value,(move[0],move[1])

			# if value > cut_value:
			# 	return cut_value,(move[0],move[1])
			if max_value < value:
				max_value = value
				max_idx = (mov[0],mov[1])

		return max_value,max_idx





	def Min(self,board,move,depth,cut_value):
		# self.count += 1
		# if self.count % 100 == 0:
		# 	print(self.count)
		my_board = copy.deepcopy(board)
		base_value = 0
		if move is not None:
			my_board.draw_xy(move[0],move[1],self.color)
			base_value = evaluate_2(my_board,self.color,move[0],move[1])
			
			if my_board.anyone_win(move[0],move[1]) == self.color:
				self.over = True
				return 9999999 * (depth + 1),(move[0],move[1])
			if my_board.anyone_win(move[0],move[1]) == self.other:
				self.over = True
				return -9999999 * (depth + 1),(move[0],move[1])
		empty_item = my_board.get_k_dist_empty(3)
		min_value = 999999999
		t = random.randint(0,len(empty_item)-1)
		min_idx = empty_item[t]
		depth -= 1
		#print('MIN_depth:' + str(depth))
		if depth < 0:
			return evaluate_2(my_board,self.color,move[0],move[1]),(move[0],move[1])

		# if base_value < cut_value:
		# 	return cut_value,(move[0],move[1])
		for mov in empty_item:
			v,idx = self.Max(my_board,mov,depth,min_value)
			value = base_value + v
			if value < cut_value:
				return cut_value,(move[0],move[1])

			if min_value > value:
				min_value = value
				min_idx = (mov[0],mov[1])

		return min_value,min_idx





class MinMax_best():
	def __init__(self,is_black,depth = 2):
		self.depth = depth
		if is_black == True:
			self.color = BLACK
			self.other = WHITE
		else:
			self.color = WHITE
			self.other = BLACK



	def get_move(self,board):
		self.over = False
		depth = self.depth
		self.count = 0
		start = time.time()
		value,idx = self.Max(board,None,depth,999999999)
		print("Time of this step: " + str(time.time() - start))
		print('AI_MINMAX had come a over:' + str(self.over))
		#print("The value is : " + str(value))
		return idx

	def Max(self,board,move,depth,cut_value):
		# self.count += 1
		# if self.count % 100 == 0:
		# 	print(self.count)
		my_board = copy.deepcopy(board)

		if move is not None:
			my_board.draw_xy(move[0],move[1],self.other)
			if my_board.anyone_win(move[0],move[1]) == self.color:
				self.over = True
				return 9999999 * (depth + 1),(move[0],move[1])
			if my_board.anyone_win(move[0],move[1]) == self.other:
				self.over = True
				return -9999999 * (depth + 1),(move[0],move[1])

		max_value = -999999999
		empty_item = my_board.get_k_dist_empty(3)
		if len(empty_item) == 0:
			empty_item.append([MIDDLE,MIDDLE])
		t = random.randint(0,len(empty_item)-1)
		max_idx = empty_item[t]
		depth -= 1
		#print('good location:  \n'+str(empty_item))
		#print('MAX_depth:'+str(depth))
		if depth < 0:
			return -evaluate(my_board,self.other),(move[0],move[1])

		
		for mov in empty_item:
			
			value,idx = self.Min(my_board,mov,depth,max_value)
			if value > cut_value:
				return cut_value,(move[0],move[1])
			if max_value < value:
				max_value = value
				max_idx = (mov[0],mov[1])

		return max_value,max_idx





	def Min(self,board,move,depth,cut_value):
		# self.count += 1
		# if self.count % 100 == 0:
		# 	print(self.count)
		my_board = copy.deepcopy(board)
		if move is not None:
			my_board.draw_xy(move[0],move[1],self.color)
			if my_board.anyone_win(move[0],move[1]) == self.color:
				self.over = True
				return 9999999 * (depth + 1),(move[0],move[1])
			if my_board.anyone_win(move[0],move[1]) == self.other:
				self.over = True
				return -9999999 * (depth + 1),(move[0],move[1])

		min_value = 999999999
		empty_item = my_board.get_k_dist_empty(3)
		if len(empty_item) == 0:
			empty_item.append([MIDDLE,MIDDLE])
		t = random.randint(0,len(empty_item)-1)
		min_idx = empty_item[t]
		depth -= 1
		#print('MIN_depth:' + str(depth))
		if depth < 0:
			return evaluate(my_board,self.color),(move[0],move[1])


		for mov in empty_item:
			value,idx = self.Max(my_board,mov,depth,min_value)
			if value < cut_value:
				return cut_value,(move[0],move[1])

			if min_value > value:
				min_value = value
				min_idx = (mov[0],mov[1])

		return min_value,min_idx

File: CPPVersion/AI.cpp
#include"AI.h"

using namespace std;
//*******************************************************************
//*                       Method of AI                              *
//*******************************************************************
AI::AI(bool is_black, int depth, int cut_v = 10) {
	if (is_black) {
		color = BLACK;
		other = WHITE;
	}
	else {
		color = WHITE;
		other = BLACK;
	}
	my_depth = depth;
	cut = cut_v;
}
void AI::set(int depth, int cut_v = -1) {
	my_depth = depth;
	cut = cut_v;
}


int* AI::AI::get_move(chessboard board) {
	count = 0;
	int move[2] = { -1,-1 };
    v_idx tmp;
    tmp = Max(board, move, my_depth, 999999999);
//    if(my_depth <= 2){
//        tmp = Max(board, move, my_depth, 999999999);
//    }
//    else{
//        tmp = Max(board, move, my_depth*2, 999999999,true);
//        if(tmp.value < 9000000 && tmp.value > -9000000){
//            tmp = Max(board, move, my_depth, 999999999);
//        }
//    }

	move[0] = tmp.i;
	move[1] = tmp.j;
	static int move_r[2] = { 0,0 };
	move_r[0] = move[0];
	move_r[1] = move[1];
	return move_r;
}
v_idx AI::Max(chessboard board, int move[2], int depth, int cut_value,bool kill) {
	v_idx tmp;
	count++;

	tmp.i = move[0];
	tmp.j = move[1];
    int winner = -1;
	if (move[0] != -1) {
		board.draw(move[0], move[1], other);
        winner = board.anyone_win(move[0],move[1]);
        if (winner == color) {
			tmp.value = 9999999 * (depth + 1);
			return tmp;
		}
        if (winner == other) {
			tmp.value = -9999999 * (depth + 1);
			return tmp;
		}
        if(winner == FULL){
            tmp.value = 0;
            return tmp;
        }
	}

    vector<Point> empty_item = board.get_k_dist_empty(3, color, cut,kill);
    if(empty_item.size() == 0 && kill){
        tmp.value = 0;

        return tmp;
    }
	if (empty_item.size() == 0) {
		Point point;
		point.i = 7;
		point.j = 7;
		empty_item.push_back(point);
	}
	v_idx max_v_idx;
	max_v_idx.i = 0;
	max_v_idx.j = 0;
	max_v_idx.value = -999999999;
	depth--;
	if (depth < 0) {
		tmp.value = evaluate(board, color);
		return tmp;
	}
	v_idx result;
	int mov[2] = { 0,0 };
	for (int c = 0; c < empty_item.size(); c++) {
		mov[0] = empty_item[c].i;
		mov[1] = empty_item[c].j;
        result = Min(board, mov, depth, max_v_idx.value,kill);
		if (result.value > cut_value) {
			tmp.value = cut_value;
			return tmp;
		}
		if (max_v_idx.value < result.value) {
			max_v_idx.value = result.value;
			max_v_idx.i = mov[0];
			max_v_idx.j = mov[1];
		}

	}
	return max_v_idx;
}


v_idx AI::Min(chessboard board, int move[2], int depth, int cut_value,bool kill) {
	v_idx tmp;
	count++;

	tmp.i = move[0];
	tmp.j = move[1];
    int winner = -1;
	if (move[0] != -1) {
		board.draw(move[0], move[1], color);
        winner = board.anyone_win(move[0],move[1]);
        if (winner == color) {
            tmp.value = 9999999 * (depth + 1);
            return tmp;
        }
        if (winner == other) {
            tmp.value = -9999999 * (depth + 1);
            return tmp;
        }
        if(winner == FULL){
            tmp.value = 0;
            return tmp;
        }
	}
    vector<Point> empty_item = board.get_k_dist_empty(3, other, cut,kill);
    if(empty_item.size() == 0 && kill){
        tmp.value = 0;

        return tmp;
    }
	if (empty_item.size() == 0) {
		Point point;
		point.i = 7;
		point.j = 7;
		empty_item.push_back(point);
	}
	v_idx min_v_idx;
	min_v_idx.i = 0;
	min_v_idx.j = 0;
	min_v_idx.value = 999999999;
	depth--;
	if (depth < 0) {
		tmp.value = evaluate(board, color);
		return tmp;
	}
	v_idx result;
	int mov[2] = { 0,0 };
	for (int c = 0; c < empty_item.size(); c++) {
		mov[0] = empty_item[c].i;
		mov[1] = empty_item[c].j;
        result = Max(board, mov, depth, min_v_idx.value,kill);
		if (result.value < cut_value) {
			tmp.value = cut_value;
			return tmp;
		}
		if (min_v_idx.value > result.value) {
			min_v_idx.value = result.value;
			min_v_idx.i = mov[0];
			min_v_idx.j = mov[1];
		}

	}
	return min_v_idx;
}
int AI::evaluate(chessboard board, int color) {
	return board.value_judge_best(color);
}
